### PR TITLE
simplify resize deciders

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change pulls up checks and removes code duplication to make those
two deciders slightly cheaper.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java [hunk 1]
```diff
@@ -90,15 +90,13 @@ public FilterAllocationDecider(Settings settings, ClusterSettings clusterSetting
...
             // only for unassigned - we filter allocation right after the index creation (for shard shrinking) to ensure
             // that once it has been allocated post API the replicas can be allocated elsewhere without user interaction
             // this is a setting that can only be set within the system!
-            IndexMetadata indexMd = allocation.metadata().getIndexSafe(shardRouting.index());
-            DiscoveryNodeFilters initialRecoveryFilters = DiscoveryNodeFilters.trimTier(indexMd.getInitialRecoveryFilters());
-            if (initialRecoveryFilters != null
-                && shardRouting.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS
-                && initialRecoveryFilters.match(node.node()) == false) {
+            IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
+            DiscoveryNodeFilters initialRecoveryFilters = DiscoveryNodeFilters.trimTier(indexMetadata.getInitialRecoveryFilters());
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java:98]: No need to load index metadata and initial recovery filters if this is not a local shards recovery

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ResizeAllocationDecider.java [hunk 2]
```diff
@@ -31,16 +30,15 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocat
...
             final IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
-            Index resizeSourceIndex = indexMetadata.getResizeSourceIndex();
+            final Index resizeSourceIndex = indexMetadata.getResizeSourceIndex();
             assert resizeSourceIndex != null;
-            if (allocation.metadata().index(resizeSourceIndex) == null) {
+            final IndexMetadata sourceIndexMetadata = allocation.metadata().index(resizeSourceIndex);
+            if (sourceIndexMetadata == null) {
                 return allocation.decision(Decision.NO, NAME, "resize source index [%s] doesn't exists", resizeSourceIndex.toString());
             }
-            IndexMetadata sourceIndexMetadata = allocation.metadata().getIndexSafe(resizeSourceIndex);
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ResizeAllocationDecider.java:43]: The result of the above null check could be reused

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ResizeAllocationDecider.java [hunk 3]
```diff
@@ -31,16 +30,15 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocat
 
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
-        final UnassignedInfo unassignedInfo = shardRouting.unassignedInfo();
-        if (unassignedInfo != null && shardRouting.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
+        if (shardRouting.unassigned() && shardRouting.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ResizeAllocationDecider.java:33]: Just looking into whether this is equivalent, I'm unsure whether it's possible to have a `ShardRouting` which is assigned but which has `unassignedInfo != null`? We check `assert (state == ShardRoutingState.UNASSIGNED && unassignedInfo == null) == false` in its constructor but maybe we should `assert (state == ShardRoutingState.UNASSIGNED) == (unassignedInfo != null)` instead?

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I think we could slightly strengthen the assertion in the constructor of `ShardRouting` to validate the changes.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**